torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,113 @@
+            inputs and labels.
+        max_seq_len (int): Maximum number of tokens to pack
+        max_rows (Optional[int]): maximum number of samples to pack. Default is None, which will pack as many samples as possible.
+        split_samples (bool): if the last sample in a pack does not fit in ``max_seq_len``,
+            split the sample into the next pack, or move it to the beginning of the next pack.
+            For pre-training, typically this is set to True for general text completion. For
+            fine-tuning, typically this is set to False to avoid truncating sentences in instruct
+            tuning. Default is False.
+    """
+
+    def __init__(
+        self,
+        ds: Dataset,
+        max_seq_len: int,
+        max_rows: Optional[int] = None,
+        split_samples: bool = False,
+    ) -> None:
+        self.ds = ds
+        self.max_seq_len = max_seq_len
+        self.max_rows = max_rows
+        self.split_samples = split_samples
+        # where final samples will be held
+        self.samples: List[Dict[str, List[int]]] = []
+        self._pack()
+
+    def _pack(self) -> None:
+        """
+        Iterate through the dataset. Use a buffer to hold samples until max_seq_len,
+        then append the buffer to self.samples as a single "packed" sample. Continue
+        until max_rows or end of dataset.
+        """
+        # buffer to hold samples until they are long enough to be added to self.samples
+        current_pack = {
+            "input_ids": [],
+            "labels": [],
+        }
+        # Keep track of what index the previous sample ends in case we need
+        # to end a pack early
+        previous_sample_boundary = 0
+
+        for input_ids, labels in tqdm(
+            self.ds, desc="Packing dataset", dynamic_ncols=True
+        ):
+            # If the dataset outputs samples that are larger than the specified
+            # max_seq_len and we're unable to split it, user needs to modify
+            # one of the two parameters
+            if len(input_ids) > self.max_seq_len and not self.split_samples:
+                raise ValueError(
+                    f"Dataset sample is too long ({len(input_ids)} > {self.max_seq_len}). "
+                    "Please set `split_samples=True` or increase `max_seq_len`."
+                )
+
+            current_pack["input_ids"].extend(input_ids)
+            current_pack["labels"].extend(labels)
+
+            if len(current_pack["input_ids"]) > self.max_seq_len:
+                current_pack = self._add_pack(
+                    current_pack=current_pack,
+                    boundary=self.max_seq_len
+                    if self.split_samples
+                    else previous_sample_boundary,
+                )
+
+            previous_sample_boundary = len(current_pack["input_ids"])
+            if self.max_rows is not None and len(self.samples) >= self.max_rows:
```

Comment:
So max_rows is only respected if we have < max_seq_len tokens right? Then why have that option at all? User can just reduce the seq len to get fewer samples? Or why would they want fewer samples?

Reply:
not exactly - max rows lets users limit how many total packs in the dataset is returned. I imagine this will be a lot more relevant for streamed / iterable datasets. max_seq_len controls the size of each individual pack. reducing max_seq_len but keeping dataset size the same will actually result in more samples.

Reply:
Yes, I agree with @RdoubleA. Max rows is how many packed rows you end with. Most users will choose `None` for datasets that fit in RAM, but will choose some lesser value for OOM datasets. A very nice helper arg might be to calculate `max_rows` for you. Not needed for now.

Reply:
Oh hmm, then I'm not sure this is clear in the doc string

> max_rows (Optional[int]): maximum number of samples to pack. Default is None, which will pack as many samples as possible.

This leads me to believe this param controls the number of samples read in not the number f packs returned? Or am I misunderstanding?

Reply:
I wrote that comment.

The intent is to signal that this arg controls the total amount of packed samples you will end up with at the end of the process. However, if you don't understand it, then others wont, either. So, the phrasing should probably be revised for clarity.